src/symbolize/gimli.rs

```diff
@@ -223,153 +245,303 @@ cfg_if::cfg_if! {
-                    for _ in 0..(*header).ncmds {
-                        segments.extend(segment(next_cmd));
-                        next_cmd = (next_cmd as usize + (*next_cmd).cmdsize as usize) as *const _;
+            let mut first_text = 0;
+            let mut text_fileoff_zero = false;
+            while let Some(cmd) = load_commands.next().ok()? {
+                if let Some((seg, _)) = cmd.segment_32().ok()? {
+                    if seg.name() == b"__TEXT" {
+                        first_text = segments.len();
+                        if seg.fileoff(endian) == 0 && seg.filesize(endian) > 0 {
+                            text_fileoff_zero = true;
+                        }
                     }
+                    segments.push(LibrarySegment {
+                        len: seg.vmsize(endian).try_into().ok()?,
+                        stated_virtual_memory_address: seg.vmaddr(endian).try_into().ok()?,
+                    });
                 }
-                libc::MH_MAGIC_64 => {
-                    let header = header as *const libc::mach_header_64;
-                    let mut next_cmd = header.offset(1) as *const libc::load_command;
-                    for _ in 0..(*header).ncmds {
-                        segments.extend(segment(next_cmd));
-                        next_cmd = (next_cmd as usize + (*next_cmd).cmdsize as usize) as *const _;
+                if let Some((seg, _)) = cmd.segment_64().ok()? {
+                    if seg.name() == b"__TEXT" {
+                        first_text = segments.len();
+                        if seg.fileoff(endian) == 0 && seg.filesize(endian) > 0 {
+                            text_fileoff_zero = true;
+                        }
                     }
+                    segments.push(LibrarySegment {
+                        len: seg.vmsize(endian).try_into().ok()?,
+                        stated_virtual_memory_address: seg.vmaddr(endian).try_into().ok()?,
+                    });
                 }
-                _ => return None,
             }
+
+            // Determine the "slide" for this library which ends up being the
+            // bias we use to figure out where in memory objects are loaded.
+            // This is a bit of a weird computation though and is the result of
+            // trying a few things in the wild and seeing what sticks.
+            //
+            // The general idea is that the `bias` plus a segment's
+            // `stated_virtual_memory_address` is going to be where in the
+            // actual address space the segment resides. The other thing we rely
+            // on though is that a real address minus the `bias` is the index to
+            // look up in the symbol table and debuginfo.
+            //
+            // It turns out, though, that for system loaded libraries these
+            // calculations are correct. For native executables, however, it
+            // appears correct. Lifting some logic from LLDB's source it has
```

Comment:
Should one of these `correct` be `incorrect`?

Reply:
Oops, yes indeed.